src/Filters/remez_fir.jl

```diff
@@ -0,0 +1,825 @@
+    weight = convert(Vector{Float64}, weight)
+    
+    grid, des, wt = build_grid(numtaps, bands, desired, weight, grid_density, filter_type)
+
+    nfilt = numtaps
+    ngrid = length(grid)
+    
+    #for j = 1 : ngrid
+    #    printfmtln("  j={}: grid[j]={}", j, grid[j])
+    #end
+    
+    neg = filter_type != filter_type_bandpass      # boolean: "neg" means negative symmetry.
+    nodd = isodd(nfilt)   # boolean: "nodd" means filter length is odd
+    nfcns = numtaps ÷ 2   # integer divide
+    if nodd && !neg
+        nfcns = nfcns + 1
+    end
+    temp = (ngrid-1) / nfcns
+    dimsize = Int64(ceil(numtaps/2.0 + 2))
+    #printfmtln("  dimsize={}   nfcns={}", dimsize, nfcns)
+
+    """/*
+     * SET UP A NEW APPROXIMATION PROBLEM WHICH IS EQUIVALENT
+     * TO THE ORIGINAL PROBLEM
+     */"""
+    if !neg
+        if !nodd
+            for j = 1 : ngrid
+                change = cospi(grid[j])
+                des[j] = des[j] / change
+                wt[j]  = wt[j] * change
+            end
+        end
+    else
+        if !nodd
+            for j = 1 : ngrid
+                change = sinpi(grid[j])
+                des[j] = des[j] / change
+                wt[j]  = wt[j]  * change
+            end
+        else
+            for j = 1 : ngrid
+                change = sinpi(2grid[j])
+                des[j] = des[j] / change
+                wt[j]  = wt[j]  * change
+            end
+        end
+    end
+        
+    iext = zeros(Int64, dimsize)   # indices of extremals
+    x = zeros(Float64, dimsize)
+    y = zeros(Float64, dimsize)
+
+    for j = 1 : nfcns
+        iext[j] = Int64(floor((j-1)*temp)) + 1
+    end
+    iext[nfcns+1] = ngrid
+
+    dev = 0.0     # deviation from the desired function, 
+                  # that is, the amount of "ripple" on the extremal set
+    devl = -1.0   # deviation on last iteration
+    nz  = nfcns+1
+    nzz = nfcns+2
+    niter = 0
+    ad = zeros(Float64, nz)
+    
+    jet = ((nfcns-1) ÷ 15) + 1
+
+    while true
+    
+        #
+        # Start next iteration
+        #        
+      @label L100
+        iext[nzz] = ngrid + 1
+        niter += 1
+        if niter > maxiter
+            break
+        end
+        
+        x[1:nz] = cospi.(2grid[iext[1:nz]])
+        #for j = 1 : nz
+        #    printfmtln("  j={}: iext[j]={} grid[iext[j]]={} x[j]={}", j, iext[j], grid[iext[j]], x[j])
+        #end
+        
+        for j = 1 : nz
+            ad[j] = lagrange_interp(j, nz, jet, x)
+        end
+
+        dnum = 0.0
+        dden = 0.0
+        k = 1
+        for j = 1 : nz
+            l = iext[j]
+            dnum += ad[j] * des[l]
+            dden += k * ad[j] / wt[l]
+            k = -k
+        end
+        dev = dnum / dden
+        # printfmtln("DEVIATION = {}", dev)
+
+        y = 0*x
+        nu, dev = initialize_y(dev, nz, iext, des, wt, y)
+        
+        if dev <= devl
+            # finished 
+            return -1  # error - deviation should always increase
```

Comment:
I think we should either throw an exception or return `nothing`. Where I'd prefer the former.